parla/task_runtime.py

```diff
@@ -327,20 +313,32 @@ def req(self):
     def req(self, new_req):
         self._req = new_req
 
+    def _wait_for_predecessor_events(self, env: TaskEnvironment):
+        if len(self._predecessor_events) > 0:
+            # Only wait events if any remaining dependent tasks exist.
+            # TODO(lhc): This does not support nested CPU tasks from GPU tasks.
+            #            When we notify successors, we don't know places on
+            #            which the successors will run since those would decide
+            #            at the next step, mapping step.
+            #            Therefore, the current runtime does not consider
+            #            successors' placements, but just creates events
+            #            on the devices where the current task is running.
+            #            For example, when CPU tasks get GPU events,
+            #            they will just skip and will not wait for them.
+            #            This is not technical limitation, just need refactoring
+            #            and adding more features.
+            #            But our target applications do not have this pattern
+            #            and will do it later.
+            env.wait_dependent_events(self._predecessor_events)
+            # Already waited all dependent events.
+            # Initialize the dependent event list.
+        self._predecessor_events = []
+
+    # Deprecated: use predecessors
     @property
     def dependencies(self) -> Tuple["Task"]:
-        with self._mutex:
-            return self._dependencies
-
-    @property
-    def dependees(self) -> Tuple["Task"]:
-        """
-        A tuple of the currently known tasks that depend on self.
-
-        This tuple may be added to at any time during the life of a task
-        (as dependee tasks are created), but tasks are never removed.
-        """
-        return tuple(self._dependees)
+        with self._mutex: # TODO(bozhi) Why?
```

Comment:
idk haha. We have too many mutexes in some places and too few in others. We certainly need to analyze each one at some point (soon)